src/lib/homestore_backend/heap_chunk_selector.cpp

```diff
@@ -59,282 +49,256 @@ csharedChunk HeapChunkSelector::select_chunk(homestore::blk_count_t count, const
-            pg_heap->available_blk_count += vchunk.available_blks();
+
+    auto pg_chunk_collection = pg_it->second;
+    auto& pg_chunks = pg_chunk_collection->m_pg_chunks;
+    std::scoped_lock lock(pg_chunk_collection->mtx);
+
+    for (size_t v_chunk_id = 0; v_chunk_id < pg_chunks.size(); ++v_chunk_id) {
+        auto chunk = pg_chunks[v_chunk_id];
+        pg_chunk_collection->m_total_blks += chunk->get_total_blks();
+        if (excluding_v_chunk_ids.find(v_chunk_id) == excluding_v_chunk_ids.end()) {
+            chunk->m_state = ChunkState::AVAILABLE;
+            ++pg_chunk_collection->available_num_chunks;
+            pg_chunk_collection->available_blk_count += chunk->available_blks();
+
+        } else {
+            chunk->m_state = ChunkState::INUSE;
         }
     }
+    return true;
 }
 
-std::shared_ptr< const std::vector <homestore::chunk_num_t> > HeapChunkSelector::get_pg_chunks(pg_id_t pg_id) const {
+std::shared_ptr< const std::vector< homestore::chunk_num_t > > HeapChunkSelector::get_pg_chunks(pg_id_t pg_id) const {
     std::shared_lock lock_guard(m_chunk_selector_mtx);
-    auto it = m_v2r_chunk_map.find(pg_id);
-    if (it != m_v2r_chunk_map.end()) {
-        return it->second;
-    } else {
+    auto pg_it = m_per_pg_chunks.find(pg_id);
+    if (pg_it == m_per_pg_chunks.end()) {
         LOGWARNMOD(homeobject, "PG {} had never been created", pg_id);
         return nullptr;
     }
+
+    auto pg_chunk_collection = pg_it->second;
+    auto& pg_chunks = pg_chunk_collection->m_pg_chunks;
+    std::scoped_lock lock(pg_chunk_collection->mtx);
+    auto p_chunk_ids = std::make_shared< std::vector< homestore::chunk_num_t > >();
+    p_chunk_ids->reserve(pg_chunks.size());
+    for (auto chunk : pg_chunks) {
+        p_chunk_ids->emplace_back(chunk->get_chunk_id());
+    }
+    return p_chunk_ids;
 }
 
-homestore::blk_alloc_hints HeapChunkSelector::chunk_to_hints(chunk_num_t chunk_id) const {
-    auto iter = m_chunks.find(chunk_id);
-    if (iter == m_chunks.end()) {
-        LOGWARNMOD(homeobject, "No chunk found for chunk_id {}, will return default blk alloc hints", chunk_id);
-        return homestore::blk_alloc_hints();
+std::optional< homestore::chunk_num_t > HeapChunkSelector::get_most_available_blk_chunk(pg_id_t pg_id) const {
+    std::shared_lock lock_guard(m_chunk_selector_mtx);
+    auto pg_it = m_per_pg_chunks.find(pg_id);
+    if (pg_it == m_per_pg_chunks.end()) {
+        LOGWARNMOD(homeobject, "No pg found for pg_id {}", pg_id);
+        return std::nullopt;
+    }
+    if (pg_it->second->available_num_chunks == 0) {
+        LOGWARNMOD(homeobject, "No available chunk for pg {}", pg_id);
+        return std::nullopt;
     }
-    homestore::blk_alloc_hints hints;
-    hints.pdev_id_hint = VChunk(iter->second).get_pdev_id();
-    return hints;
+
+    std::scoped_lock lock(pg_it->second->mtx);
+    auto pg_chunk_collection = pg_it->second;
+    auto& pg_chunks = pg_chunk_collection->m_pg_chunks;
+    auto max_it =
```

Comment:
is there a case that no chunk is avaiable ? should we handle this case?

Reply:
I think it worth check that max_it is pointing to a available chunk

Reply:
pls if you can add into your new pr

Reply:
ok，its a more stright way, i'll change into that way with new pr.